### DISCUSSION TITLE
Very long tracebacks

### DISCUSSION THREAD

**Discussion by hhoeflin** [Textualize/rich#1258] [Q&A]
Hi,

I am getting very long tracebacks in "locals". This is despite max_string and max_length. Looking at the code, it looks like only strings are being truncated, no other objects. The representation of other objects is used as is (see pretty.traverse) and with some objects these can be very long (e.g. json objects).

Does anyone have a suggestion on how to deal with it, or should this be a feature request (e.g. to limit string-representations of objects also to the max_string length.

Thanks!

**Comment by willmcgugan** [comment]
What do you mean re "json object"? If it's a string, it will be truncated. If it's a dict / list then the number of elements will be truncated to `max_length`.

I'm not convinced it would be a good idea to truncate other object's `__repr__`. If the output is too long it should probably be up to the developer how to limit the size.

**Reply by hhoeflin** [reply]
Thanks for answering so quickly! I guess it depends by what you mean by "max_length" in the case of nested objects. In my case, I have geojson objects. The number of items in the outer dict is not large, but there can be thousands of coordinates - I have outputs in the locals-traceback that are hundreds of lines long for a single one of the geojson items. The object is from the `geojson` package, and I don't have control over it. 

An example below. I truncated after a few lines, but this output goes on for about 500 lines.

```
│ │           res = {                                                                            │ │  
│ │                 │   'WF': {"features": [{"geometry": {"coordinates": [[[3761.455897,         │ │  
│ │                 1248.961812], [3756.447301, 1234.391349], [3741.876839, 1214.812291],        │ │  
│ │                 [3737.095906, 1210.031358], [3722.29778, 1190.452299], [3712.735914,         │ │  
│ │                 1170.87324], [3702.946385, 1156.302778], [3678.586393, 1131.942787],         │ │  
│ │                 [3663.788267, 1117.144661], [3659.007334, 1107.582795], [3644.436872,        │ │  
│ │                 1092.784669], [3629.638746, 1078.214207], [3619.849217, 1068.424678],        │ │  
│ │                 [3605.278755, 1058.635148], [3595.489226, 1048.845619], [3576.13783,         │ │  
│ │                 1024.485627], [3566.348301, 1009.915165], [3556.558772, 995.344703],         │ │  
│ │                 [3546.769242, 990.336107], [3536.979713, 975.765644], [3532.19878,           │ │  
│ │                 965.976115], [3517.400654, 946.397056], [3502.830192, 922.037065],           │ │  
│ │                 [3498.049259, 917.256132], [3488.25973, 897.677073], [3478.470201,           │ │  
│ │                 883.106611], [3463.899738, 863.527552], [3449.101613, 843.948494],           │ │  
│ │                 [3439.539747, 829.378031], [3429.750217, 814.807569], [3414.952092,          │ │  
│ │                 805.01804], [3405.390226, 790.447577], [3385.811167, 770.868519],            │ │  
│ │                 [3376.021638, 761.078989], [3361.451176, 746.508527], [3346.65305,           │ │  
│ │                 731.938065], [3332.082588, 717.139939], [3307.722596, 697.788544],           │ │  
│ │                 [3293.152134, 682.990418], [3278.354008, 668.419956], [3263.783546,          │ │  
│ │                 653.849494], [3253.994016, 639.051368], [3239.423554, 624.480906],           │ │  
│ │                 [3224.853092, 614.691376], [3205.274033, 600.120914], [3195.484504,          │ │  
│ │                 585.550452], [3180.914042, 575.760923], [3171.124512, 561.19046],            │ │  
│ │                 [3156.55405, 551.400931], [3146.764521, 536.602805], [3127.185462,           │ │  
│ │                 522.032343], [3112.615, 507.461881], [3102.82547, 497.672351], [3088.255008, │ │  
│ │                 483.101889], [3073.456882, 473.31236], [3054.105487, 458.741898],            │ │  
│ │                 [3034.526429, 448.952368], [3019.955966, 439.162839], [3000.376908,          │ │  
│ │                 429.37331], [2980.797849, 419.58378], [2966.227387, 409.794251],             │ │  
│ │                 [2941.867395, 395.223789], [2922.288337, 385.434259], [2912.498807,          │ │  
│ │                 380.653326], [2897.928345, 370.863797], [2888.138816, 370.863797],           │ │  
│ │                 [2873.568353, 365.855201], [2858.997891, 361.074268], [2844.199765,          │ │  
│ │                 361.074268], [2839.418832, 361.074268], [2834.410236, 361.074268],           │ │  
│ │                 [2824.84837, 365.855201], [2819.839774, 365.855201], [2815.058841,           │ │  
│ │                 380.653326], [2805.269312, 390.442856], [2800.260715, 400.004722],           │ │  
│ │                 [2795.479782, 414.802847], [2790.698849, 424.592377], [2780.90932,           │ │  
│ │                 439.162839], [2771.119791, 453.733301], [2766.111194, 458.741898],           │ │  
│ │                 [2761.330261, 468.303763], [2761.330261, 473.31236], [2756.549328,           │ │  
│ │                 483.101889], [2756.549328, 487.882822], [2751.540732, 492.891419],           │ │  
│ │                 [2746.759799, 507.461881], [2746.759799, 512.242814], [2741.751203,          │ │  
│ │                 522.032343], [2736.97027, 531.821872], [2736.97027, 541.611402],             │ │  
│ │                 [2736.97027, 551.400931], [2736.97027, 561.19046], [2736.97027, 565.971393], │ │  
│ │                 [2746.759799, 575.760923], [2751.540732, 580.541856], [2761.330261,          │ │  
│ │                 590.331385], [2775.900724, 600.120914], [2785.690253, 609.910444],           │ │  
│ │                 [2795.479782, 619.699973], [2805.269312, 629.489502], [2810.050244,          │ │  
│ │                 639.051368], [2819.839774, 653.849494], [2834.410236, 668.419956],           │ │  
│ │                 [2834.410236, 678.209485], [2844.199765, 687.999015], [2844.199765,          │ │  
│ │                 692.779948], [2849.208362, 707.35041], [2849.208362, 717.139939],            │ │  
│ │                 [2853.989295, 731.938065], [2853.989295, 741.499931], [2853.989295,          │ │  
│ │                 770.868519], [2849.208362, 795.22851], [2839.418832, 814.807569],            │ │  
│ │                 [2834.410236, 834.386628], [2829.629303, 848.95709], [2815.058841,           │ │  
│ │                 873.317082], [2805.269312, 892.89614], [2790.698849, 912.247535],            │ │  
│ │                 [2775.900724, 931.826594], [2756.549328, 946.397056], [2741.751203,          │ │  
│ │                 961.195182], [2722.399807, 975.765644], [2707.601682, 990.336107],           │ │  
```

**Reply by willmcgugan** [reply]
The locals in Traceback should be limited to 10 items, but that doesn't appear the case in your example. It's also not formatted in the Rich way.

I suspect that your local isn't actually a dictionary, it may be an object that works like a dictionary, but is some other type. 

Could you `print(type(res))`. If I'm correct, then that won't say `dict`. 

Also please make sure you have the latest version, there was some recent changes to pretty printing.

**Reply by willmcgugan** [reply]
It might be `res["WF"]` that isn't a real dict.

**Comment by hhoeflin** [comment]
yes - that isn't a real dict. It is a geojson.FeatureCollection object.

**Reply by willmcgugan** [reply]
Ah, so in that case Rich will just call the `__repr__` which is returning a very long string.

It would be possible for Rich to truncate those with an ellipsis, but the result would not be as readable as other containers.

There is a protocol in Rich where a special method can be added to allow pretty printing of unknown object types, but it's not fully fleshed out yet. Once that's ready you could perhaps convince the geo Django devs to add it. It wouldn't even require Rich as a dependency.

**Comment by hhoeflin** [comment]
and I have the latest version installed.

**Comment by hhoeflin** [comment]
Ok, thanks for looking into it. Alternatively could the pretty-printing
module allow for a function to be passed in that handle the printing for
certain objects? e.g., a function for printing all objects inheriting from
an arbitrary base class? Truncating long representations with an ellipsis
would then be a special case of this (where you pass a method that handles
printing for all objects inheriting from "Object" and just cut off the
result. Truncating strings at max_length would also be a special case.

But in any case - thanks so much for this library. It is super useful as it
is.

**Comment by hhoeflin** [comment]
Just wanted to add to this, that similar problems occur for numpy arrays, for pytorch tensors, for pytorch models etc. Long '__repr__' for objects are relatively common. 

I had to switch off the "locals" again, as it just was not practical.
